src/server/unit_sao.cpp

```diff
@@ -183,79 +220,67 @@ void UnitSAO::getAttachment(int *parent_id, std::string *bone, v3f *position,
 	*force_visible = m_force_visible;
 }
 
+void UnitSAO::clearAnyAttachments()
+{
+	// This is called before this SAO is marked for removal/deletion and unlinks
+	// any parent or child relationships.
+	// This is done at this point and not in ~UnitSAO() so attachments to
+	// "phantom objects" don't stay around while we're waiting to be actually deleted.
+	// (which can take several server steps)
+	clearParentAttachment();
+	clearChildAttachments();
+}
+
 void UnitSAO::clearChildAttachments()
 {
 	// Cannot use for-loop here: setAttachment() modifies 'm_attachment_child_ids'!
 	while (!m_attachment_child_ids.empty()) {
-		int child_id = *m_attachment_child_ids.begin();
+		const auto child_id = *m_attachment_child_ids.begin();
 
 		// Child can be NULL if it was deleted earlier
-		if (ServerActiveObject *child = m_env->getActiveObject(child_id))
-			child->setAttachment(0, "", v3f(0, 0, 0), v3f(0, 0, 0), false);
-
-		removeAttachmentChild(child_id);
-	}
-}
-
-void UnitSAO::clearParentAttachment()
-{
-	ServerActiveObject *parent = nullptr;
-	if (m_attachment_parent_id) {
-		parent = m_env->getActiveObject(m_attachment_parent_id);
-		setAttachment(0, "", m_attachment_position, m_attachment_rotation, false);
-	} else {
-		setAttachment(0, "", v3f(0, 0, 0), v3f(0, 0, 0), false);
+		if (auto *child = m_env->getActiveObject(child_id))
+			child->clearParentAttachment();
```

Comment:
When `child == nullptr`, `m_attachment_child_ids` will never shrink. Endless loop.

Reply:
If we get into that situation that's generally a problem but I'll handle it here at least

Reply:
Granted, I did provoke this and the other comment at the same time with the following recursive snippet:
```Lua
	on_punch = function(self, puncher, _, _, _, _)
		self.object:set_detach()
		error("oops")
	end,
	on_detach = function(self, parent)
		self.object:set_attach(parent, "", {x=30,y=0,z=30}, {x=45,y=45,z=0})
		print("on_detach")
	end,
```
where `self.object` is attached to another object, and you hit it. Used this testing code: https://github.com/SmallJoker/minetest_lab/blob/master/init_18_attachment_chain.lua